### DISCUSSION TITLE
Experimental partition API arguments

### DISCUSSION THREAD

**Discussion by abkfenris** [dagster-io/dagster#4729] [Q&A] [answered]
I'm trying to use the new experimental partition APIs, but it appears that some of the arguments that the old APIs include aren't carried forwards, which makes it harder to adapt the partitions to different timing.

Specifically, I'm trying to use [@monthly_partitioned_config](https://docs.dagster.io/_apidocs/experimental#dagster.monthly_partitioned_config), but I cannot access the current month. With the [previous API](https://docs.dagster.io/_apidocs/schedules-sensors#dagster.monthly_schedule) you could set `partition_months_offset=0` to get the current month.

Setup:

```py
@graph
def example_graph():
    pass

@monthly_partitioned_config(start_date="2021-08-01")
def monthly_config(start, _end):
    return {}

job = example_graph.to_job(config=monthly_config)

mode = job.get_default_mode()
partition_def = mode.get_partition_set_def(job.name)
```

If I try to access the current month, I get an exception.

```py
>>> partition = partition_def.get_partition(name="2021-09-01")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.9/site-packages/dagster/core/definitions/partition.py", line 411, in get_partition
    check.failed("Partition name {} not found!".format(name))
  File "/opt/conda/lib/python3.9/site-packages/dagster/check/__init__.py", line 119, in failed
    raise CheckError("Failure condition: {desc}".format(desc=desc))
dagster.check.CheckError: Failure condition: Partition name 2021-09-01 not found!
```

If I try to access the previous month, then it works fine:
```py
>>> partition = partition_def.get_partition(name="2021-08-01")
>>> partition
<dagster.core.definitions.partition.Partition object at 0x7f5f9106f280>
```

In this case I'm using the partition with a sensor instead of as a sensor as I don't control the upstream data, instead I'm triggering as data becomes available and aggregating all of the currently available for the current month.

Is there a way to adapt the experimental partition APIs to allow access to the current month? (or future! I may have some forecasting data I'm working with soon that can 'backfill' dates ahead...)

This is with 0.12.8.
```py
>>> dagster.__version__
'0.12.8'
```

**Comment by sryza** [comment, marked as answer]
@abkfenris - do your partitions represent time windows?  If so, does the time window for the partition that you want to be named "2021-08-01" end on 8/1 or begin on 8/1?

**Reply by abkfenris** [reply]
Time windows starting at the beginning of the month. 

So for today I want to generate 2021-09-01 with the two days of data available so far this month. At some point shortly after 0 UTC when the new satellite data is out, I'd like to regenerate 2021-09 with the next days worth of data to keep the monthly archive up to date.

**Reply by sryza** [reply]
Got it - makes total sense.  Would it be correct to say that you only want to apply the offset at the end of the range of partitions?

I.e., with no offset, if you specify your partitions with the following code, and if today is 2021-09-03, you'd end up with a partition set with a single partition: `[(start=2021-08-01, end=2021-09-01)]`.
```
@monthly_partitioned_config(start_date="2021-08-01")
def my_config(start, end):
    ...
```

I would expect that a parameter called "partition_months_offset" set to 1 would essentially shift that to `[(start=2021-09-01, end=2021-10-01)]`.  Would it be accurate to say that you would actually want two partitions?  I.e. `[(start=2021-08-01, end=2021-09-01), (start=2021-09-01, end=2021-10-01)]`?  So maybe we need something like `end_offset`?

**Reply by abkfenris** [reply]
Yes, I do want two partitions, so something like `end_offset` would make sense. I hadn't thought about the partition end times. 

```py
@monthly_partitioned_config(start_date="2021-08-01", end_offset=1)
def my_config(start, end):
     ...
```
To give `[(start=2021-08-01, end=2021-09-01), (start=2021-09-01, end=2021-10-01)]`

If I was working with forecasts of climatology for the next 10 years:

```py
@monthly_partitioned_config(start_date="2021-08-01", end_offset=((12 * 10) + 1))
def my_config(start, end):
     ...
```
To give `[(start=2021-08-01, end=2021-09-01), (start=2021-09-01, end=2021-10-01), ... (start=2031-08-01, end=2031-09-01), (start=2031-09-01, end=2031-10-01)]`

---

Might it make it more sense to have it inclusive vs exclusive, and have the default then be `-1`?

```py
@monthly_partitioned_config(start_date="2021-08-01", end_offset=0)
def my_config(start, end):
     ...
```
To give `[(start=2021-08-01, end=2021-09-01), (start=2021-09-01, end=2021-10-01)]`

10 year forecast

```py
@monthly_partitioned_config(start_date="2021-08-01", end_offset=12 * 10)
def my_config(start, end):
     ...
```
To give `[(start=2021-08-01, end=2021-09-01), (start=2021-09-01, end=2021-10-01), ... (start=2031-08-01, end=2031-09-01), (start=2031-09-01, end=2031-10-01)]`

**Reply by sryza** [reply]
Got it - makes total sense.  I'll look into adding an `end_offset` argument.

> Might it make it more sense to have it inclusive vs exclusive, and have the default then be -1?

I don't think it would be crazy, but I think that, in general, zero is the expected default for offset params, so would prefer to keep it at that.  For the 10 year forecast, you could do `12 * 10 + 1` if I'm not mistaken.

**Reply by abkfenris** [reply]
Awesome, thank you.

**Reply by sryza** [reply]
Here's a diff: https://github.com/dagster-io/dagster/pull/4741/files

**Reply by abkfenris** [reply]
That looks great. The updated doc strings help make it much more clear that a partition is a window, not just a single time.
